daemon/algod/api/server/v2/dryrun.go

```diff
@@ -0,0 +1,522 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package v2
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/daemon/algod/api/server/v2/generated"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// DryrunApp holds global app state for a dryrun call.
+// TODO: should become obsolete when app supported added to api v2?
+type DryrunApp struct {
+	Creator  basics.Address   `codec:"a"`
+	AppIndex uint64           `codec:"i"`
+	Params   basics.AppParams `codec:"p"`
+}
+
+// DryrunLocalAppState holds per-account app state for a dryrun call.
+// TODO: should become obsolete when app supported added to api v2?
+type DryrunLocalAppState struct {
+	Account  basics.Address       `codec:"a"`
+	AppIndex basics.AppIndex      `codec:"i"`
+	State    basics.AppLocalState `codec:"s"`
+}
+
+// DryrunSource is TEAL source text that gets uploaded, compiled, and inserted into transactions or application state
+type DryrunSource struct {
+	// FieldName is what kind of sources this is.
+	// If lsig then it goes into the transactions[this.TxnIndex].LogicSig
+	// If approv or clearp it goes into the Approval Program or Clear State Program of application[this.AppIndex]
+	FieldName string `codec:"f"` // "lsig", "approv", "clearp"
+	Source    string `codec:"text"`
+	TxnIndex  int    `codec:"ti,omitempty"`
+	AppIndex  uint64 `codec:"ai,omitempty"`
+}
+
+// DryrunRequest is the JSON object uploaded to /v2/transactions/dryrun
+// Given the Transactions and simulated ledger state upload, run TEAL scripts and return debugging information.
+type DryrunRequest struct {
+	// Txns is transactions to simulate
+	Txns []transactions.SignedTxn `codec:"txns,omitempty"`
+
+	// Accounts
+	// Optional, useful for testing Application Call txns.
+	Accounts []generated.Account `codec:"accounts,omitempty"`
+
+	Apps []DryrunApp `codec:"apps,omitempty"`
+
+	AccountAppStates []DryrunLocalAppState `codec:"applocal,omitempty"`
+
```

Comment:
Why not []basics.AccountData?

Reply:
DryrunLocalAppState contains a v2 API Account object instead of the internal AccountData. I'm hoping that will be easier to use for API users.